aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -563,6 +564,67 @@ Tensor masked_fill(const Tensor & self, const Tensor & mask, const Tensor & sour
   return result;
 }
 
+static Tensor & masked_select_out_impl_cpu(Tensor & result, const Tensor & self, const Tensor & mask) {
+  NoNamesGuard guard;
+
+  TORCH_CHECK(mask.scalar_type() == ScalarType::Byte || mask.scalar_type() == ScalarType::Bool,
+              "masked_select: expected BoolTensor or ByteTensor for mask");
+  TORCH_CHECK(self.scalar_type() == result.scalar_type(),
+              "masked_select(): self and result must have the same scalar type");
+
+  Tensor _mask, _self;
+  std::tie(_mask, _self) = expand_outplace(mask, self);
+
+  auto shape = _self.sizes().vec();
+  auto mask_long = at::empty(shape, self.options().dtype(at::kLong)).copy_(_mask);
+  int64_t numel = mask_long.sum().item().toLong();
+  result.resize_({numel});
+  if (numel == 0) {
+    return result;
+  }
+
+  // Create strided view of result before feeding into TensorIterator
+  auto strides = DimVector(shape.size(), 0);
+  auto result_strided = result.as_strided(shape, strides);
+
+  // Use a prefix sum to record the output locations of the masked elements,
+  // so as to parallel with TensorIterator.
+  auto mask_prefix_sum = at::empty(shape, self.options().dtype(at::kLong));
+  auto mask_long_data = mask_long.data_ptr<int64_t>();
+  auto mask_prefix_sum_data = mask_prefix_sum.data_ptr<int64_t>();
+  // TODO: Here can only use std::partial_sum for C++14,
+  // use std::exclusive_scan when PyTorch upgrades to C++17, which have better peformance.
+  // std::exclusive_scan(mask_long_data, mask_long_data + mask_long.numel(), mask_prefix_sum_data, 0);
+  std::partial_sum(mask_long_data, mask_long_data + mask_long.numel(), mask_prefix_sum_data);
```

Comment:
`cumsum` now runs in sequential so performance not very good. I have another colleague working on `cumsum` (see #33280 ), it's almost finished but even with that `cumsum` is still sequential on a flat tensor with `dim=0`.

`std::partial_sum` from STL has slightly better perf on CPU than `cumsum`. Anyway, the diff is not very big, if you feel like a unified code path for both CUDA and CPU, i assume use `cumsum` is also acceptable. But I was wondering that [thrust::exclusive_scan](https://docs.nvidia.com/cuda/thrust/index.html#prefix-sums) should promise a good performance on CUDA, is there a problem with that?

```bash
#use std::partial_sum
input size: [128 1000]; output size: [63941]: time = 0.295 ms
input size: [256 1000]; output size: [127483]: time = 0.413 ms
input size: [512 1000]; output size: [255929]: time = 0.700 ms
input size: [1024 1000]; output size: [511179]: time = 1.238 ms

# use cumsum
input size: [128 1000]; output size: [63941]: time = 0.330 ms
input size: [256 1000]; output size: [127483]: time = 0.482 ms
input size: [512 1000]; output size: [255929]: time = 0.729 ms
input size: [1024 1000]; output size: [511179]: time = 1.283 ms
```

Reply:
If the extra bit of performance that `std::partial_sum` gives is important to you, then it's alright with me. Maybe someone from Facebook should weigh in though.

However, it seems like it would be ideal if we could make cumsum use `std::partial_sum` underneath so that it gives the similar performance to what you have. Also, do you happen to know what the performance would be if you use @XiaobingSuper 's updated `cumsum`?

Reply:
That makes sense. I think you can submit a new PR with an unified CPU and CUDA implementation to replace this one :)